sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala

```diff
@@ -62,20 +61,113 @@ case class CustomShuffleReaderExec private(
     }
   }
 
-  override def stringArgs: Iterator[Any] = Iterator(description)
+  override def stringArgs: Iterator[Any] = {
+    val desc = if (isLocalReader) {
+      "local"
+    } else {
+      if (hasCoalescedPartition && hasSkewedPartition) {
+        "coalesced and skewed"
+      } else if (hasCoalescedPartition) {
+        "coalesced"
+      } else if (hasSkewedPartition) {
+        "skewed"
+      } else {
+        ""
+      }
+    }
+    Iterator(desc)
+  }
 
-  private var cachedShuffleRDD: RDD[InternalRow] = null
+  def hasCoalescedPartition: Boolean = {
+    partitionSpecs.exists(_.isInstanceOf[CoalescedPartitionSpec])
+  }
 
-  override protected def doExecute(): RDD[InternalRow] = {
-    if (cachedShuffleRDD == null) {
-      cachedShuffleRDD = child match {
-        case stage: ShuffleQueryStageExec =>
-          new ShuffledRowRDD(
-            stage.shuffle.shuffleDependency, stage.shuffle.readMetrics, partitionSpecs.toArray)
-        case _ =>
-          throw new IllegalStateException("operating on canonicalization plan")
+  def hasSkewedPartition: Boolean = {
+    partitionSpecs.exists(_.isInstanceOf[PartialReducerPartitionSpec])
+  }
+
+  def isLocalReader: Boolean = {
+    if (partitionSpecs.exists(_.isInstanceOf[PartialMapperPartitionSpec])) {
+      assert(partitionSpecs.forall(_.isInstanceOf[PartialMapperPartitionSpec]))
+      true
+    } else {
+      false
+    }
+  }
+
+  private def shuffleStage = child match {
+    case stage: ShuffleQueryStageExec => Some(stage)
+    case _ => None
+  }
+
+  override lazy val metrics = {
```

Comment:
nit: use variables like "partitionSizeMetrics", "skewPartitionMetrics" to make the flow a little clearer?